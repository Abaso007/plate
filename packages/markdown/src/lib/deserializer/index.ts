/**
 * @file Automatically generated by barrelsby.
 */

export * from './convertChildrenDeserialize';
export * from './convertNodesDeserialize';
export * from './convertTextsDeserialize';
export * from './deserializeMd';
export * from './mdastToSlate';
export * from './utils/index';
