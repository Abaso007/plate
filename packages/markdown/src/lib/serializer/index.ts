export * from './convertNodes';
export * from './convertTexts';
export * from './serializeMd';
export * from './serializeInlineMd';
export * from './types';
