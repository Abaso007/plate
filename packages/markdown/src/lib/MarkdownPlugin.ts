import type { Plugin } from 'unified';

import {
  type OmitFirst,
  type PluginConfig,
  bindFirst,
  createTSlatePlugin,
  isUrl,
  KEYS,
} from '@udecode/plate';

import type { MdRules, PlateType } from './types';

import { deserializeInlineMd, deserializeMd } from './deserializer';
import { serializeMd } from './serializer';

export type AllowNodeConfig = {
  /** Custom filter function for nodes during deserialization */
  deserialize?: (node: any) => boolean;
  /** Custom filter function for nodes during serialization */
  serialize?: (node: any) => boolean;
};

export type MarkdownConfig = PluginConfig<
  'markdown',
  {
    /**
     * Configuration for allowed node types. Cannot be combined with
     * disallowedNodes.
     */
    allowedNodes: PlateType[] | null;
    /**
     * Configuration for disallowed node types. Cannot be combined with
     * allowedNodes.
     *
     * @default null
     */
    disallowedNodes: PlateType[] | null;
    /**
     * Array of remark plugins to extend Markdown parsing and serialization
     * functionality. For example, you can add remark-gfm to support GFM syntax,
     * remark-math to support mathematical formulas, etc. These plugins will be
     * used during the parsing and generation of Markdown text.
     *
     * @default undefined
     */
    remarkPlugins: Plugin[];
    /**
     * Rules that define how to convert Markdown syntax elements to Slate editor
     * elements. Or rules that how to convert Slate editor elements to Markdown
     * syntax elements. Includes conversion rules for elements such as
     * paragraphs, headings, lists, links, images, etc.
     *
     * You can pass null disable default node parser.
     *
     * NOTE: don't forget pass `mark:true` when you custom inline nodes.
     *
     * @default null
     */
    rules: MdRules | null;
    /**
     * Custom filter function for nodes during deserialization and
     * serialization.
     *
     * @default null
     */
    allowNode?: AllowNodeConfig;
  },
  {
    markdown: {
      deserialize: OmitFirst<typeof deserializeMd>;
      deserializeInline: OmitFirst<typeof deserializeInlineMd>;
      serialize: OmitFirst<typeof serializeMd>;
    };
  }
>;

export const MarkdownPlugin = createTSlatePlugin<MarkdownConfig>({
  key: KEYS.markdown,
  options: {
    allowedNodes: null,
    disallowedNodes: null,
    remarkPlugins: [],
    rules: null,
  },
})
  .extendApi(({ editor }) => ({
    deserialize: bindFirst(deserializeMd, editor),
    deserializeInline: bindFirst(deserializeInlineMd, editor),
    serialize: bindFirst(serializeMd, editor),
  }))
  .extend(({ api }) => ({
    parser: {
      format: 'text/plain',
      deserialize: ({ data }) => api.markdown.deserialize(data),
      query: ({ data, dataTransfer }) => {
        const htmlData = dataTransfer.getData('text/html');

        if (htmlData) return false;

        const { files } = dataTransfer;

        if (
          !files?.length && // if content is simply a URL pass through to not break LinkPlugin
          isUrl(data)
        ) {
          return false;
        }

        return true;
      },
    },
  }));
