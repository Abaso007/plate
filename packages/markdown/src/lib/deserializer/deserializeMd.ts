import type { Descendant, SlateEditor } from '@udecode/plate';
import type { Root } from 'mdast';
import type { Plugin } from 'unified';

import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import remarkMdx from 'remark-mdx';
import remarkParse from 'remark-parse';
import { unified } from 'unified';

import type { TNodes } from '../nodesRule';

import { MarkdownPlugin } from '../MarkdownPlugin';
import { mdastToSlate } from './mdastToSlate';


// TODO: fixes tests
const remarkSplitLineBreaks: Plugin<[deserializeOptions?], Root, Root> = function ({ splitLineBreaks = false } = {}) {
  if (!splitLineBreaks) return;

  return function transformer(tree: Root) {
    const processNode = (node: any) => {
      if (!node.position) return node;

      if (node.children) {
        const newChildren: any[] = [];
        let lastLine = node.position.start.line;

        node.children.forEach((child: any, index: number) => {
          const isFirstChild = index === 0;
          const isLastChild = index === node.children.length - 1;

          // Add empty paragraphs for line breaks before the child
          const emptyLinesBefore = child.position.start.line - (isFirstChild ? lastLine : lastLine + 1);
          if (emptyLinesBefore > 0) {
            for (let i = 0; i < emptyLinesBefore; i++) {
              newChildren.push({
                children: [{ type: 'text', value: '' }],
                position: {
                  end: { column: 1, line: lastLine + i + 1 },
                  start: { column: 1, line: lastLine + i + 1 }
                },
                type: 'paragraph'
              });
            }
          }

          // Add the child node
          newChildren.push(processNode(child));

          // Add empty paragraphs for line breaks after the last child
          if (isLastChild) {
            const emptyLinesAfter = node.position.end.line - child.position.end.line - 1;
            if (emptyLinesAfter > 0) {
              for (let i = 0; i < emptyLinesAfter; i++) {
                newChildren.push({
                  children: [{ type: 'text', value: '' }],
                  position: {
                    end: { column: 1, line: child.position.end.line + i + 1 },
                    start: { column: 1, line: child.position.end.line + i + 1 }
                  },
                  type: 'paragraph'
                });
              }
            }
          }

          lastLine = child.position.end.line;
        });

        node.children = newChildren;
      }

      return node;
    };

    return processNode(tree);
  };
};

export const deserializeMd = (editor: SlateEditor, data: string) => {
  // if using remarkMdx, we need to replace <br> with <br /> since <br /> is not supported in mdx.
  data = data.replaceAll('<br>', '<br />');

  const { nodes, splitLineBreaks } = editor.getOptions(MarkdownPlugin);

  const toSlateProcessor = unified()
    .use(remarkParse)
    .use(remarkMdx)
    .use(remarkGfm)
    .use(remarkMath)
    .use(remarkSplitLineBreaks, { splitLineBreaks })
    .use(remarkToSlate, {
      nodes,
      splitLineBreaks,
    });

  return toSlateProcessor.processSync(data).result;
};

export type deserializeOptions = {
  nodes?: TNodes;
  splitLineBreaks?: boolean;
};

declare module 'unified' {
  interface CompileResultMap {
    remarkToSlateNode: Descendant[];
  }
}

const remarkToSlate: Plugin<[deserializeOptions?], Root, Descendant[]> =
  function ({ nodes, splitLineBreaks = false } = {}) {
    this.compiler = function (node) {
      return mdastToSlate(node as Root, { nodes, splitLineBreaks });
    };
  };
