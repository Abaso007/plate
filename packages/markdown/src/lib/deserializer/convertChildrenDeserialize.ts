import type { Descendant } from '@udecode/plate';

import type { MdRootContent } from '../mdast';
import type { MdDecoration } from '../types';
import type { DeserializeMdOptions } from './deserializeMd';

import { convertNodesDeserialize } from './convertNodesDeserialize';

export const convertChildrenDeserialize = (
  children: MdRootContent[],
  deco: MdDecoration,
  options: DeserializeMdOptions
): Descendant[] => {
  if (children.length === 0) {
    return [{ text: '' }];
  }

  return convertNodesDeserialize(children, deco, options);
};
