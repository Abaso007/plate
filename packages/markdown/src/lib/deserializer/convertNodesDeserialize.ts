import type { Descendant } from '@udecode/plate';

import type { MdRootContent } from '../mdast';
import type { deserializeOptions } from './deserializeMd';
import type { Decoration } from './type';

import { defaultNodes } from '../nodesRule';
import { getPlateNodeType } from '../utils';
import { convertChildren } from './covertChildren';

export const convertNodesDeserialize = (
  nodes: MdRootContent[],
  deco: Decoration,
  options: deserializeOptions
): Descendant[] => {
  return nodes.reduce<Descendant[]>((acc, node) => {
    acc.push(...buildSlateNode(node, deco, options));
    return acc;
  }, []);
};

export const buildSlateNode = (
  mdastNode: MdRootContent,
  deco: Decoration,
  options: deserializeOptions
): Descendant[] => {
  const optionNodes = options.nodes;

  /** Handle custom mdx nodes */
  if (mdastNode.type === 'mdxJsxTextElement') {
    if (mdastNode.name === 'br') {
      const parserKey = mdastNode.name;
      const nodeParserDeserialize =
        optionNodes?.[parserKey]?.deserialize ??
        defaultNodes[parserKey]?.deserialize;

      if (nodeParserDeserialize)
        return nodeParserDeserialize(mdastNode, deco, options);

      return [{ text: '\n' }];
    }

    if (mdastNode.name === 'u') {
      const parserKey = 'underline';

      const nodeParserDeserialize =
        optionNodes?.[parserKey]?.deserialize ??
        defaultNodes[parserKey]?.deserialize;

      if (nodeParserDeserialize)
        return nodeParserDeserialize(mdastNode, deco, options) as any;

      return convertChildren(mdastNode.children, { underline: true }, options);
    }
  }

  const type = getPlateNodeType(mdastNode.type);

  const nodeParser =
    options.nodes?.[type]?.deserialize ?? defaultNodes[type]?.deserialize;

  if (nodeParser) {
    const result = nodeParser(mdastNode, deco, options);
    return Array.isArray(result) ? result : [result];
  }
  return [];
};
