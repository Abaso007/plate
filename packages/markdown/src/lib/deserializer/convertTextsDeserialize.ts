import type { MdDelete, MdEmphasis, MdStrong } from '../mdast';
import type { MdDecoration } from '../types';
import type { DeserializeMdOptions } from './deserializeMd';

import { mdastToPlate } from '../types';
import { buildSlateNode } from './convertNodesDeserialize';

export const convertTextsDeserialize = (
  mdastNode: MdDelete | MdEmphasis | MdStrong,
  deco: MdDecoration,
  options: DeserializeMdOptions
) => {
  return mdastNode.children.reduce((acc: any, n: any) => {
    acc.push(
      ...buildSlateNode(
        n,
        { ...deco, [mdastToPlate(options.editor!, mdastNode.type)]: true },
        options
      )
    );
    return acc;
  }, []);
};
