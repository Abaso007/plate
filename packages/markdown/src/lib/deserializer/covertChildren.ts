import type { Descendant } from '@udecode/plate';

import type { MdRootContent } from '../mdast';
import type { deserializeOptions } from './deserializeMd';
import type { Decoration } from './type';

import { convertNodesDeserialize } from './convertNodesDeserialize';

export const convertChildren = (
  children: MdRootContent[],
  deco: Decoration,
  options: deserializeOptions
): Descendant[] => {
  if (children.length === 0) {
    return [{ text: '' }];
  }

  return convertNodesDeserialize(children, deco, options);
};
