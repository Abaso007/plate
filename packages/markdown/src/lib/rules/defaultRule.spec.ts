import { H1Plugin } from '@platejs/basic-nodes/react';
import { createPlateEditor, ParagraphPlugin } from 'platejs/react';

import { deserializeMd } from '../deserializer';
import { MarkdownPlugin } from '../MarkdownPlugin';
import { serializeMd } from '../serializer';

describe('defaultRules', () => {
  it('should serialize default keys', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'h1',
      },
      {
        children: [{ text: 'Paragraph' }],
        type: 'p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [MarkdownPlugin, H1Plugin, ParagraphPlugin],
    });

    const result = serializeMd(editor, { value: nodes });
    expect(result).toMatchSnapshot();
  });

  it('should serialize custom keys', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'custom-h1',
      },
      {
        children: [{ text: 'Paragraph' }],
        type: 'custom-p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [
        MarkdownPlugin,
        H1Plugin.configure({
          node: { type: 'custom-h1' },
        }),
        ParagraphPlugin.configure({
          node: { type: 'custom-p' },
        }),
      ],
    });

    const result = serializeMd(editor, { value: nodes });
    expect(result).toMatchSnapshot();
  });

  it('should deserialize default keys', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'h1',
      },
      {
        children: [{ text: 'Paragraph' }],
        type: 'p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [MarkdownPlugin, H1Plugin, ParagraphPlugin],
    });

    const result = deserializeMd(editor, '# Heading 1\nParagraph');
    expect(result).toEqual(nodes);
  });

  it('should deserialize custom keys', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'custom-h1',
      },
      {
        children: [{ text: 'Paragraph' }],
        type: 'custom-p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [
        MarkdownPlugin,
        H1Plugin.configure({
          node: { type: 'custom-h1' },
        }),
        ParagraphPlugin.configure({
          node: { type: 'custom-p' },
        }),
      ],
    });

    const result = deserializeMd(editor, '# Heading 1\nParagraph');
    expect(result).toEqual(nodes);
  });
});
