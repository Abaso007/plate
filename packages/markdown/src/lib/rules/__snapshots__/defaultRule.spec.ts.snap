// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultRules should serialize custom keys 1`] = `
"# Heading 1

Paragraph
"
`;

exports[`defaultRules should serialize default keys 1`] = `
"# Heading 1

Paragraph
"
`;
